### ISSUE TITLE
private$predict_filter stores a reference to the training data - this is a issue when saving the fitted object.

### ISSUE THREAD

**Issue by AndreasNordland** [kkholst/targeted#226]
```
## targeted::learner retains the training data after $estimate()
##
## Reprex for an issue to file against the `targeted` package.
##
## Observation
## -----------
## After calling `learner$estimate(data)`, the fitted learner object retains a
## reference to the training `data` in the environment of the internal
## `private$predict_filter` function. Serializing / saving the fitted learner
## therefore embeds a full copy of the training design matrix, which can
## dominate the size of the fitted object (especially for lightweight
## backends such as glm or xgboost).
##
## This is an issue for downstream packages that store one fitted learner per
## stage and per cross-fit fold (e.g. polle), because the training data ends
## up duplicated many times in the resulting model object.
##
## The naive workaround -- store only the bare fit and call
## `unfitted_clone$predict(newdata, object = fit)` on a clean clone of the
## learner prototype -- does not work either: `private$predict_filter` is
## NULL on a learner that has not been `$estimate()`-d, so `$predict()`
## errors with "attempt to apply non-function".
##
## Environment
## -----------
## Tested with:
##   R 4.5.2, targeted 0.9.0
##
## Suggested fix
## -------------
## Either
##   (a) initialise `private$predict_filter` to an identity function at
##       construction time, so that `$predict(newdata, object = fit)` works
##       on a learner that has not been $estimate()-d locally, or
##   (b) avoid capturing the training `data` in the closure environment of
##       the predict filter (e.g. compute whatever the filter needs from
##       `data` up front and drop the reference).
```

stopifnot(requireNamespace("targeted", quietly = TRUE))
library(targeted)
cat("targeted version:", format(packageVersion("targeted")), "\n\n")

set.seed(1)
n <- 5000
d <- data.frame(
  y  = rnorm(n),
  x1 = rnorm(n),
  x2 = rnorm(n),
  x3 = rnorm(n)
)
cat(sprintf("training data:      %6.3f MB (n = %d)\n\n",
            length(serialize(d, NULL)) / 1e6, n))

## Helper: is the training `data` reachable from the fitted learner?
reachable_data <- function(l) {
  e <- environment(l$predict.filter)
  if (is.null(e)) return(NULL)
  if (!"data" %in% ls(e, all.names = TRUE)) return(NULL)
  get("data", envir = e)
}

## ---------------------------------------------------------------------------

show_retention <- function(make_learner, label) {
  l <- make_learner()
  l$estimate(d)

  before <- length(serialize(l, NULL)) / 1e6
  pred_before <- head(l$predict(d), 5)

  ## Try to clear the retained data. This reaches into targeted internals and
  ## is only used here to demonstrate the source of the overhead.
  captured <- reachable_data(l)
  if (!is.null(captured)) {
    assign("data", NULL, envir = environment(l$predict.filter))
  }

  after <- length(serialize(l, NULL)) / 1e6
  pred_after <- head(l$predict(d), 5)

  cat(sprintf(
    "%-16s  size: %6.3f MB (fitted)  ->  %6.3f MB (data cleared)   %s\n",
    label, before, after,
    if (isTRUE(all.equal(pred_before, pred_after)))
      "predictions unchanged"
    else
      "PREDICTIONS DIFFER (!)"
  ))
  cat(sprintf(
    "%-16s  training data reachable via environment(l$predict.filter)$data: %s\n",
    "", !is.null(captured)
  ))
  invisible(NULL)
}

cat("--- fitted learner size, before/after clearing $predict.filter env ---\n")
show_retention(function() learner_glm(y ~ ., family = gaussian()), "learner_glm")

if (requireNamespace("xgboost", quietly = TRUE)) {
  show_retention(function() learner_xgboost(y ~ ., nrounds = 3),
                 "learner_xgboost")
}

show_retention(
  function() learner_sl(
    list(a = learner_glm(y ~ .), b = learner_glm(y ~ 1)),
    nfolds = 2
  ),
  "learner_sl"
)

##
## does not work.

cat("\n--- storing only the bare fit and predicting via an unfitted clone ---\n")

proto <- learner_glm(y ~ ., family = gaussian())
fit_learner <- proto$clone(deep = TRUE)
fit <- fit_learner$estimate(d)          # the bare underlying fit object

unfitted_clone <- proto$clone(deep = TRUE)
res <- tryCatch(
  head(unfitted_clone$predict(d, object = fit), 3),
  error = function(e) e
)
cat(sprintf(
  "unfitted_clone$predict(newdata, object = fit): %s\n",
  if (inherits(res, "error")) paste0("ERROR: ", conditionMessage(res))
  else paste(format(res), collapse = " ")
))

## came is itself fitted (so that predict_filter is populated):
cat(sprintf(
  "fit_learner$predict(newdata, object = fit):    %s\n",
  paste(format(head(fit_learner$predict(d, object = fit), 3)), collapse = " ")
))

## ---------------------------------------------------------------------------

cat("\n--- source of the retention ---\n")
l <- learner_glm(y ~ ., family = gaussian())
l$estimate(d)
env <- environment(l$predict.filter)
cat("names in environment(l$predict.filter):\n  ",
    paste(ls(env, all.names = TRUE), collapse = ", "), "\n", sep = "")
cat("identical(env$data, training d): ", identical(env$data, d), "\n", sep = "")
